File: TooManyReviewers/TableViewModel.swift
//
//  TableViewModel.swift
//  TooManyReviewers
//
//  Created by MeowCat on 2025/2/1.
//

import SwiftUI

class TableViewModel: ObservableObject {
    typealias Pos = Pair<Int, Int>
    
    enum PathSegmentKind {
        enum LinearDirection {
            case horizontal, vertical
        }
        
        enum EndDirection {
            case up, down, left, right
        }
        
        enum TurnDirection {
            case upLeft, upRight, downLeft, downRight
        }
        
        case linear(_: LinearDirection)
        case end(_: EndDirection)
        case turn(_: TurnDirection)
    }
    
    @Published
    var items: [[[Item]]] = []
    
    @Published
    var pathMask: [[PathSegmentKind?]]?
    
    var rowIndices: [Int] { Array(0..<items.count) }
    
    var colIndices: [Int] { items.isEmpty ? [] : Array(0..<items[0].count) }
    
    @Published
    var selected: Pos?
    
    private func reset(items: [[[Item]]]) {
        self.items = items
        pathMask = nil
        selected = nil
    }
    
    func placeholders(
        numRows: Int, numCols: Int
    ) {
        reset(items: (1...numRows).map { i in
            (1...numCols).map { j in
                [.textual("\(i)\(j)")]
            }
        })
    }
    
    func generate(numRows: Int, numCols: Int) {
        let builder = TableBuilder(
            numRows: numRows, numCols: numCols)
        builder.build()
        let fn = SimpleCache { (id: Int) -> TableViewModel.Item in
            return Bool.random() ?
                .textual(NameGenerator.randomAbbr)
            : .image(Self.icon(index: Int.random(in: 0..<Self.iconSheetCount)))
        }
        reset(items: builder.grid.map { row in
            row.map { ids in ids.map { fn[$0] } }
        })
    }
    
    static let iconSheet = NSImage(named: "Avatars")?
        .cgImage(forProposedRect: nil, context: nil, hints: nil)
    
    static let iconSheetRowCapacity = 12
    
    static let iconSheetCount = 128
    
    static let iconSize = 128
    
    static func icon(index: Int) -> NSImage? {
        let (y, x) = index.quotientAndRemainder(dividingBy: iconSheetRowCapacity)
        guard let image = iconSheet?.cropping(to: CGRect(
            x: x * iconSize,
            y: y * iconSize,
            width: iconSize,
            height: iconSize)) else { return nil }
        return NSImage(cgImage: image, size: NSSize(width: iconSize, height: iconSize))
    }
    
    func handleClick(row: Int, col: Int) {
        let current = Pair(row, col)
        if let selected, selected != current {
            let removes = Set(items[row][col]).intersection(items[selected.first][selected.second])
            if !removes.isEmpty {
                let path = Set(path(from: selected, to: current))
                if !path.isEmpty {
                    for pos in [selected, current] {
                        items[pos.first][pos.second] = items[pos.first][pos.second]
                            .filter { !removes.contains($0) }
                    }
                    pathMask = rowIndices.map { row in
                        colIndices.map { col in
                            let pos = Pos(row, col)
                            let neighbors: Int = Pos.fourDirections.reduce(0) {
                                ($0 << 1) | (path.contains(pos + $1) ? 1 : 0)
                            }
                            if !path.contains(Pos(row, col)) { return nil }
                            return switch neighbors {
                            case 0b1000: .end(.up)
                            case 0b0100: .end(.right)
                            case 0b0010: .end(.down)
                            case 0b0001: .end(.left)
                            case 0b1100: .turn(.upRight)
                            case 0b1010: .linear(.vertical)
                            case 0b1001: .turn(.upLeft)
                            case 0b0110: .turn(.downRight)
                            case 0b0101: .linear(.horizontal)
                            case 0b0011: .turn(.downLeft)
                            default: nil
                            }
                        }
                    }
                }
            }
            self.selected = nil
        } else { selected = current }
    }
    
    func path(from fromPos: Pos, to toPos: Pos) -> [Pos] {
        let dx = abs(fromPos.first - toPos.first)
        let dy = abs(fromPos.second - toPos.second)
        if dx == 0 && dy == 1 || dx == 1 && dy == 0 {
            return [fromPos, toPos]
        }
        let d = Dijkstra(grid: items.map { row in
            row.map { $0.isEmpty ? 0 : 1 }
        })
        return d.dijkstra(start: fromPos, end: toPos)
    }
}

extension TableViewModel {
    enum Item {
        case textual(_: String)
        case image(_: NSImage?)
        
        var textualDescription: String {
            switch self {
            case .textual(let text): text
            case .image(_): "(image)"
            }
        }
    }
}

extension TableViewModel.Item: Equatable, Hashable { }

File: src/table_viewmodel.ts
import { immerable } from "immer"
import { NameGenerator } from "./name_generator";
import { Pair, Pos, Pos$add, Pos$fourDirections } from "./pair";
import { SimpleCache } from "./simple_cache";
import { TableBuilder } from "./table_builder";
import { Optional, U } from "./utils";
import { Dijkstra } from "./dijkstra";
import { OSet } from "./oset";
import { Updater } from "use-immer";

export type PathSegmentKindLinear = "horizontal" | "vertical";

export type PathSegmentKindEnd = "up" | "down" | "left" | "right";

export type PathSegmentKindTurn = "upLeft" | "upRight" | "downLeft" | "downRight";

export type PathSegmentKind = PathSegmentKindLinear | PathSegmentKindEnd | PathSegmentKindTurn;

export type TableViewModelItem = string | [number, number];

let iconSheet: Optional<ImageData>

export function updateIconSheet(imageData: ImageData) {
  iconSheet = imageData;
}

const iconSheetRowCapacity = 12;

const iconSheetCount = 128;

export class TableViewModel {
  [immerable] = true;

  public pathMask: Optional<Optional<PathSegmentKind>[][]>;

  public rowIndices: number[];

  public colIndices: number[];

  public selected: Optional<Pos>;

  constructor(
    public items: TableViewModelItem[][][],
  ) {
    this.rowIndices = [...items.keys()];
    this.colIndices = items.length == 0 ? [] : [...items[0].keys()];
  }

  static generate(
    numRows: number, numCols: number, isMeow: boolean
  ): TableViewModel {
    // let x = [...Array(numRows).keys()].map((i) =>
    //   [...Array(numCols).keys()].map((j) => {
    //     let m = (i % 4) + 1;
    //     let k = (i * numCols + j) * m;
    //     return [...Array(m).keys()].map((i) => [
    //       Math.trunc((k + i) / iconSheetRowCapacity),
    //       (k + i) % iconSheetRowCapacity,
    //     ]);
    //   }));
    // return new TableViewModel(x as any);
    let builder = new TableBuilder(numRows, numCols);
    builder.build();
    let covered = new OSet();
    const randomName = isMeow ? NameGenerator.randomKit : NameGenerator.randomAbbr;
    let fn = new SimpleCache<number, TableViewModelItem>((_) => {
      while (true) {
        const next = Math.random() < 0.5 ?
          randomName() : this.randomIcon();
        if (covered.add(next)) { return next; }
      }
    });
    return new TableViewModel(builder.grid.map((row) => {
      return row.map((ids) => {
        return ids.map((id) => fn.get(id));
      });
    }));
  }

  static randomIcon(): [number, number] {
    let index = U.randomInt(iconSheetCount);
    return [
      Math.trunc(index / iconSheetRowCapacity),
      index % iconSheetRowCapacity,
    ];
  }

  handleClick(row: number, col: number, setState: Updater<TableViewModel>) {
    let current = new Pair(row, col)
    let selected = this.selected;
    if (selected && !current.equalsTo(selected)) {
      let removes = new OSet(this.items[row][col]).intersection(
        this.items[selected.first][selected.second]);
      if (removes.size > 0) {
        let path = new OSet(this.path(selected, current));
        if (path.size > 0) {
          for (let pos of [selected, current]) {
            this.items[pos.first][pos.second] = this.items[pos.first][pos.second]
              .filter((v) => !removes.has(v));
          }
          this.pathMask = this.rowIndices.map((row) => {
            return this.colIndices.map((col) => {
              let pos = new Pair(row, col);
              let neighbors: number = Pos$fourDirections.reduce(
                (p: number, c) => (p << 1) | (path.has(Pos$add(pos, c)) ? 1 : 0),
                0);
              if (!path.has(pos)) { return undefined; }
              switch (neighbors) {
                case 0b1000: return 'up';
                case 0b0100: return 'right';
                case 0b0010: return 'down';
                case 0b0001: return 'left';
                case 0b1100: return 'upRight';
                case 0b1010: return 'vertical';
                case 0b1001: return 'upLeft';
                case 0b0110: return 'downRight';
                case 0b0101: return 'horizontal';
                case 0b0011: return 'downLeft';
                default: return undefined;
              }
            });
          });
        }
      }
      this.selected = undefined;
    } else { this.selected = current; }
    setTimeout(() => setState(
      (state) => state.pathMask = undefined
    ), 200);
  }

  public path(fromPos: Pos, toPos: Pos): Pos[] {
    let dx = Math.abs(fromPos.first - toPos.first);
    let dy = Math.abs(fromPos.second - toPos.second);
    if (dx == 0 && dy == 1 || dx == 1 && dy == 0) {
      return [fromPos, toPos];
    }
    let d = new Dijkstra(this.items.map(
      (row) => row.map((e) => Number(e.length != 0))));
    return d.dijkstra(fromPos, toPos);
  }
}
